src/twisted/conch/ssh/factory.py

```diff
@@ -31,14 +33,35 @@ class SSHFactory(protocol.Factory):
         b"ssh-connection": connection.SSHConnection,
     }
 
+    # For convenience and wider protocol compatibility, add some key variants
+    # using stronger hash algorithms.
+    _extraHashAlgorithms = [
+        (b"ssh-rsa", b"rsa-sha2-256", hashes.SHA256()),
+        (b"ssh-rsa", b"rsa-sha2-512", hashes.SHA512()),
+    ]
+
     def startFactory(self):
         """
         Check for public and private keys.
         """
         if not hasattr(self, "publicKeys"):
-            self.publicKeys = self.getPublicKeys()
+            self.publicKeys = dict(self.getPublicKeys())
+            for source, target, hashAlgorithm in self._extraHashAlgorithms:
+                if source in self.publicKeys and target not in self.publicKeys:
+                    key = keys.Key.fromString(
+                        self.publicKeys[source].toString("OPENSSH")
+                    )
+                    key.hashAlgorithm = hashAlgorithm
+                    self.publicKeys[target] = key
         if not hasattr(self, "privateKeys"):
-            self.privateKeys = self.getPrivateKeys()
+            self.privateKeys = dict(self.getPrivateKeys())
+            for source, target, hashAlgorithm in self._extraHashAlgorithms:
+                if source in self.privateKeys and target not in self.privateKeys:
+                    key = keys.Key.fromString(
+                        self.privateKeys[source].toString("OPENSSH")
+                    )
+                    key.hashAlgorithm = hashAlgorithm
```

Comment:
Since `fromString` is a constructor, I would go with something like this 

```suggestion
                    key = keys.Key.fromString(
                        self.privateKeys[source].toString("OPENSSH")
                        hashAlgorithm=hashAlgorithm,
                    )
```

but the current API is also OK and I not -1 on it.

The proposed API would meant that we don't need a setter for `hashAlgorithm` and all validation can be done at `init` or `fromString` level.

Reply:
This is now fixed by not having `hashAlgorithm` be a property of the key at all.